### ISSUE TITLE
Bitwise operators

### ISSUE THREAD

**Issue by shreyasminocha** [shreyasminocha/nirvana#8]
Do you think these are still relevant? Do you think a very high level language like Nirvana needs these today?

**Comment by rakhack** [comment]
To my understanding of your language's idea of at least, bitwise operations are too low-level. Though they might (and most likely will) get necessary when dealing with hashing and encryption stuff.

**Comment by shreyasminocha** [comment]
@rakhack I concur. Maybe they could be moved to the stdlib in some way?

**Comment by rakhack** [comment]
@shreyasminocha As functions? I'm not exactly sure there is a precedent for that. It might be an idea, though.

Just want to notice: how do you plan for `and`, `or` operators to function? For instance, in C++ the following is correct:

    1 | 2 == 3
    1 || 2 == 1

Which way do you want `or` operator to work?

**Comment by shreyasminocha** [comment]
@rakhack From what I understand, `|` is bitwise or and `||` is logical or in C++. Thus `1 || 2` performs an implicit type cast, making the statement effectively `true || true` (correct me if I'm incorrect). I'm not pro implicit typecasts. The plan for the `or` operator is for it to work only on two booleans. If we *do* decide to implement bitwise operators as part of the language, they'll work like your first example, of course.

**Comment by rakhack** [comment]
@shreyasminocha Yep. Then I suppose bitwise operators could be realized as functions (`std.bitwise.shift(int a, int s)` or something of the kind).

Or, possibly, support custom operators (like `int foo = bar std.bitwise.lshift baz;`). The downsides are:
 - it's not as readable as good ol' `|, &, ^, <<, >>, etc`,
 - goes slightly against "one way to do stuff" rule.

On the good side, such approach would cover operator overloading #7 quite nicely - no need for overloading per se, one can just easily define their own operators. And that would be a distinctive feature for sure.

**Comment by shreyasminocha** [comment]
Hmm... I don't want to rush this decision. I'll leave this open for others to weigh in. Personally, I'm kind of torn between the advantages and disadvantages...

**Comment by rakhack** [comment]
It's ultimately up to you :smiley: I'm merely suggesting.
Actually, there's one thing slightly connected to my suggestion that you might be interested in: Unified Function Call Syntax. A popular implementation is found in D language: [link 1](http://ddili.org/ders/d.en/ufcs.html), [link 2](https://tour.dlang.org/tour/en/gems/uniform-function-call-syntax-ufcs).

**Comment by shreyasminocha** [comment]
@rakhack, sure, will look. I've been told to look into D and E by some others, so I'm going to do that too.
